### ISSUE TITLE
Rename tasks when fusing

### ISSUE THREAD

**Issue by mrocklin** [dask/dask#1911]
This task runs very quickly:

```python
{'max-123': (np.max, my_array)}
```

However this task takes a long time:

```python
{'dot-456': (np.dot, x, y),
 'max-456': (np.max, 'dot-456')}

# which becomes

{'max-456': (np.max, (np.dot, x, y))
```

However both `max-123` and `max-456` are thought to be `max` tasks.  If you have done `max` operations in the past then the scheduler will probably look at your `max-456` task and say "well, this is gonna be really fast, I should definitely keep it exactly where it is" and avoid any sort of load balancing.

It would be convenient to rename tasks when fusing.  This also affects distributed diagnostics like the progressbars.

@jcrist or @eriknw do either of you have time / interest to look into this?

**Comment by eriknw** [comment]
`key_split` in `distributed.utils` determines how key names are interpreted.  Note that this is in `distributed`, not `dask`.  To keep `dask.optimize.fuse` general, perhaps it could accept an optional argument that determines the new keys for fused tasks.

**Comment by mrocklin** [comment]
Yeah, this is more of a problem whenever you don't want the scheduler to inspect the actual tasks.  The single machine schedulers are comfortable doing this but the distributed scheduler isn't.  For what it's worth in my head the `dask/distributed` repository is as much a part of dask as the single-machine schedulers, despite living in a separate repository.

In general though I agree that there is some danger here, especially for the boundary/output keys.  I think that, for as long as we're mucking around with the values of internal tasks, we also have the opportunity to muck around with the keys.  The current choice of "use the topmost fused task" is pretty reasonable, but still decently arbitrary.  Constructing a new key according to a particular scheme seems equally arbitrary.  If there are multiple schemes to support then I agree that consuming a function would be best.  If there are not multiple schemes then I would be tempted to just bake one in.

However, given that I'm asking other people to do the work, I'm more than happy to go along with things :)  It would be valuable in many ways to current dask (distributed) users if names of fused tasks were not the same as the names of their outermost subtasks.

**Comment by eriknw** [comment]
> The current choice of "use the topmost fused task" is pretty reasonable, but still decently arbitrary. Constructing a new key according to a particular scheme seems equally arbitrary.

I disagree that using the topmost fused task key is arbitrary.  Other tasks may reference the topmost tasks, so keeping them the same is most efficient choice.  To change the topmost task key efficiently, we would either need to (1) simply use an alias for the previous key, `{old_key: new_key}` (would this give the desired behavior in `distributed`?), (2) inline the `{old_key: new_key}` aliases, or (3) leverage `dependents`--which `fuse` otherwise doesn't need--to update all tasks that use the topmost key.  Options (2) and (3) result in cleaner dask graphs, but requires additional processing.

**Comment by eriknw** [comment]
For the new task key name, any preference between pipeline-style, `{'max-dot-123': (dot, (max, ...))}`, or function composition-style, `{'dot-max-123': (dot, (max, ...))}`?

**Comment by mrocklin** [comment]
Ah, I hadn't thought of dependencies at all.  

No strong thoughts on naming.  I guess I personally have a slight preference for inner-to-outer (`max-dot`) just as I find `toolz.pipe` more intuitive than `toolz.compose`.

**Comment by jcrist** [comment]
Fixed by #1919. Closing.
